test/Test-Stores.shared.js

```diff
@@ -0,0 +1,180 @@
+      this.server.datastore.should.have.property('create')
+      done()
+    })
+
+    it('must have a remove method', function (done) {
+      this.server.datastore.should.have.property('remove')
+      done()
+    })
+
+    it('must have a write method', function (done) {
+      this.server.datastore.should.have.property('write')
+      done()
+    })
+
+    it('must have a getOffset method', function (done) {
+      this.server.datastore.should.have.property('getOffset')
+      done()
+    })
+  })
+}
+
+exports.shouldCreateUploads = function () {
+  describe('create', function () {
+    const testFileSize = 960244
+    const invalidReq = { headers: {}, url: this.storePath }
+    const req = {
+      headers: {
+        'upload-length': testFileSize.toString(),
+        'upload-metadata': 'foo bar',
+      },
+      url: this.storePath,
+    }
+
+    it('should reject if both upload-length and upload-defer-length are not provided', function (done) {
+      assert.rejects(() => this.server.datastore.create(invalidReq))
+      done()
+    })
+
+    it('should reject when namingFunction is invalid', function (done) {
+      const namingFunction = (incomingReq) => incomingReq.doesnotexist.replace(/\//g, '-')
+      this.server.datastore.generateFileName = namingFunction
+      assert.rejects(() => this.server.datastore.create(req))
+      done()
+    })
+
+    it('should create a file, process it, and send file created event', async function () {
+      const fileCreatedEvent = sinon.fake()
+
+      this.server.on(EVENTS.EVENT_FILE_CREATED, fileCreatedEvent)
+
+      const file = await this.server.datastore.create(req)
+      assert.equal(file instanceof File, true)
+      assert.equal(file.upload_length, testFileSize)
+      assert.equal(fileCreatedEvent.calledOnce, true)
+    })
+
+    it('should use custom naming function when provided', async function () {
+      const namingFunction = () => 'hardcoded-name'
+
+      this.server.datastore.generateFileName = namingFunction
+
+      const file = await this.server.datastore.create(req)
+      assert.equal(file instanceof File, true)
+      assert.equal(file.id, 'hardcoded-name')
+      assert.equal(file.upload_length, testFileSize)
+    })
+  })
+}
+
+exports.shouldRemoveUploads = function () {
+  describe('remove (termination extension)', function () {
+    it('should reject when the file does not exist', function () {
+      const req = { file_id: '1234' }
+      return this.server.datastore.remove(req).should.be.rejected()
+    })
+
+    it('should delete the file when it does exist', async function () {
+      const fileDeletedEvent = sinon.fake()
+      const req = { headers: { 'upload-length': 1000 }, url: this.storePath }
+
+      this.server.datastore.on(EVENTS.EVENT_FILE_DELETED, fileDeletedEvent)
+
+      const file = await this.server.datastore.create(req)
+      await this.server.datastore.remove({ file_id: file.id })
+      assert.equal(fileDeletedEvent.calledOnce, true)
```

Comment:
This test ensures that the event is correctly fired. Is it also possible to check whether subsequent calls for the deleted file returns errors?